openforcefield/topology/topology.py

```diff
@@ -1149,6 +1156,32 @@ def box_vectors(self, box_vectors):
             assert len(box_vectors) == 3
         self._box_vectors = box_vectors
 
+    @property
+    def is_periodic(self):
+        """Return whether or not this Topology is intended to be described with periodic
+        boundary conditions."""
+        if self._is_periodic is None:
+            self._is_periodic = self.box_vectors is not None
+        return self._is_periodic
+
+    @is_periodic.setter
+    def is_periodic(self, is_periodic):
+        """
+        Set the partial charge model used for all molecules in the topology.
+
+        Parameters
+        ----------
+        is_periodic : bool
+            Whether or not this Topology is periodici
+
+        """
+        if is_periodic and self.box_vectors is None:
+            raise InvalidPeriodicityError(
+                "Cannot set is_periodic to True without box vectors. Consider setting "
+                ".box_vectors first"
+            )
+        self._is_periodic = is_periodic
```

Comment:
(blocking) Is is necessary to store `self._is_periodic` at all? The performance gain should be marginal, but the added complexity of having a data member that could fall out of sync seems pretty high.

Reply:
It depends on whether or not the difference between `None` and "haven't set it yet" is important. I'll push what it would look like and seek some more input on it.

More generally, it depends on if we want `.is_periodic`, `.box_vectors`, none, or both to live here long-term vs. other solutions, like carrying them alongside other objects in the process of system creation.

Reply:
Here's what it looks like: https://gist.github.com/mattwthompson/2e9d71e4e22cdc3f469bd948d8f01765

Reply:
Thanks for the gist. It is a bit awkward to have an error that says "Error: go unset this other thing before you set this thing", but it does keep the _actual_ complexity down to what it was before. **I'm in favor of the functionality you show there, and that you have implemented currently**

Until the `System` object work progresses to the point where we have a strong preference/spec for what's allowed further down the pipeline, let's make this PR as "minimally invasive" as possible -- it's not an API break if we ALLOW an object to be in a new state later, but it is if we DISALLOW a previously-OK state.